// genie-toolkit/dist/lib/training/base_training_job.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2019 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseTrainingJob = void 0;
const path_1 = __importDefault(require("path"));
const events = __importStar(require("events"));
class BaseTrainingJob extends events.EventEmitter {
    constructor(options) {
        var _a;
        super();
        this._killed = false;
        this.id = options.id;
        this._datadir = path_1.default.resolve(options.datadir);
        this._workdir = path_1.default.resolve(options.workdir);
        this._outputdir = options.outputdir ? path_1.default.resolve(options.outputdir) : null;
        this._debug = (_a = options.debug) !== null && _a !== void 0 ? _a : true;
        this._progress = 0;
        this.child = null;
        this.metrics = {};
    }
    get datadir() {
        return this._datadir;
    }
    get workdir() {
        return this._workdir;
    }
    get outputdir() {
        return this._outputdir;
    }
    get debug() {
        return this._debug;
    }
    get progress() {
        return this._progress;
    }
    set progress(value) {
        // progress is monotonic
        if (value > this._progress) {
            this._progress = value;
            this.emit('progress', value);
        }
    }
    kill() {
        this._killed = true;
        if (this.child)
            this.child.kill('SIGTERM');
    }
}
exports.BaseTrainingJob = BaseTrainingJob;
