// thingtalk/dist/runtime/values.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of ThingTalk
//
// Copyright 2018-2020 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Aggregation = exports.RecurrentTimeRule = exports.Currency = exports.Time = exports.Location = exports.StringLike = exports.Entity = void 0;
const assert_1 = __importDefault(require("assert"));
const polyfill_1 = require("@js-temporal/polyfill");
const escaping_1 = require("../utils/escaping");
/**
 * Runtime representation of an entity value.
 *
 */
class Entity {
    /**
     * Construct a new entity value.
     *
     * @param value - the entity identifier value
     * @param display - optional human-readable display name for the entity
     */
    constructor(id, display, options = {}) {
        this.value = id;
        this.display = display || null;
        // set the property only if provided
        // this allows subclasses to implement the method normally
        if (options.softmatch)
            this.softmatch = options.softmatch;
    }
    toString() {
        return this.value;
    }
    /**
     * Compile the entity value to JS code.
     *
     * @return {string} JS code that evaluates to this entity value
     */
    toJSSource() {
        return `new __builtin.Entity(${(0, escaping_1.stringEscape)(this.value)}, ${(0, escaping_1.stringEscape)(this.display)})`;
    }
    /**
     * Check if this JS value looks like an entity.
     *
     * @param {any} obj - the JS value to check
     * @return {boolean} `true` if the value is a string or an instance of this class
     */
    static isEntity(obj) {
        return obj instanceof Entity || typeof obj === 'string';
    }
}
exports.Entity = Entity;
/**
 * Runtime representation of a string-like object.
 *
 * Objects of this class can be used to override the behavior of
 * string operators at runtime.
 */
class StringLike {
    /**
     * Construct a new string-like object.
     *
     * @param value - the string value
     */
    constructor(id, options = {}) {
        this.value = id;
        // set the property only if provided
        // this allows subclasses to implement the method normally
        if (options.softmatch)
            this.softmatch = options.softmatch;
    }
    toString() {
        return this.value;
    }
    valueOf() {
        return this.value;
    }
    /**
     * Compile this to JS code.
     *
     * @return {string} JS code that evaluates to this value
     * @package
     */
    toJSSource() {
        return (0, escaping_1.stringEscape)(this.value);
    }
}
exports.StringLike = StringLike;
/**
 * Runtime representation of a location value.
 *
 */
class Location {
    /**
     * Construct a new location value.
     *
     * @param {number} lat - latitude in degrees (between -90 and 90)
     * @param {number} lon - longitude in degrees (between -180 and 180)
     * @param {string|null} [display] - human-readable display name for this location
     */
    constructor(lat, lon, display) {
        if (typeof lat !== 'number' || typeof lon !== 'number')
            throw new Error(`Invalid location (${lat}, ${lon})`);
        this.x = lon;
        this.y = lat;
        this.display = display || null;
    }
    /**
     * Latitude value.
     */
    get lat() {
        return this.y;
    }
    /**
     * Longitude value.
     */
    get lon() {
        return this.x;
    }
    toString() {
        if (this.display !== null)
            return this.display;
        else
            return '[Latitude: ' + Number(this.y).toFixed(5) + ' deg, Longitude: ' + Number(this.x).toFixed(5) + ' deg]';
    }
    /**
     * Compile the location value to JS code.
     *
     * @return {string} JS code that evaluates to this location value
     * @package
     */
    toJSSource() {
        return `new __builtin.Location(${this.y}, ${this.x}, ${(0, escaping_1.stringEscape)(this.display)})`;
    }
    /**
     * Check if this JS value looks like an location.
     *
     * For compatibility reasons, the runtime representation of a location is not required
     * to be an instance of the {@link Builtin.Location} class, and can be any object
     * with `x` and `y` own properties.
     *
     * @param {any} obj - the JS value to check
     * @return {boolean} `true` if the value looks like a location
     * @package
     */
    static isLocation(obj) {
        return (obj instanceof Location || (typeof obj === 'object' && obj !== null &&
            Object.prototype.hasOwnProperty.call(obj, 'x') && Object.prototype.hasOwnProperty.call(obj, 'y')));
    }
}
exports.Location = Location;
/**
 * Runtime representation of a time value.
 *
 */
class Time {
    /**
     * Construct a new time value.
     *
     * @param {number} hour - hour value (between 0 and 11)
     * @param {number} minute - minute value (between 0 and 59)
     * @param {number} [second] - second value (between 0 and 59)
     */
    constructor(hour, minute, second = 0) {
        if (!(hour >= 0) || !(minute >= 0) || !(second >= 0))
            throw new Error(`Invalid time ${hour}:${minute}:${second}`);
        this.hour = hour;
        this.minute = minute;
        this.second = second;
    }
    static fromSeconds(seconds) {
        const hour = Math.floor(seconds / 3600);
        seconds -= hour * 3600;
        const minute = Math.floor(seconds / 60);
        seconds -= minute * 60;
        return new Time(hour, minute, seconds);
    }
    static fromTemporal(time) {
        return new Time(time.hour, time.minute, time.second);
    }
    // for comparisons
    /**
     * Convert this time value to the number of seconds since midnight.
     *
     * This can be used to compare time values using `<` and `>`
     * @return {number} the number of seconds since midnight
     */
    valueOf() {
        return this.hour * 3600 + this.minute * 60 + this.second;
    }
    toString() {
        if (this.second === 0)
            return `${this.hour}:${this.minute < 10 ? '0' : ''}${this.minute}`;
        else
            return `${this.hour}:${this.minute < 10 ? '0' : ''}${this.minute}:${this.second < 10 ? '0' : ''}${this.second}`;
    }
    toJSON() {
        return this.toString();
    }
    /**
     * Compile the time value to JS code.
     *
     * @return {string} JS code that evaluates to this time value
     * @package
     */
    toJSSource() {
        return `new __builtin.Time(${this.hour}, ${this.minute}, ${this.second})`;
    }
}
exports.Time = Time;
/**
 * Runtime representation of a currency value.
 *
 */
class Currency {
    constructor(value, code) {
        this.value = value;
        this.code = code;
    }
    valueOf() {
        return this.value;
    }
    toString() {
        return `${this.value} ${this.code.toUpperCase()}`;
    }
    toLocaleString(locale, options = {}) {
        options.style = 'currency';
        options.currency = this.code.toUpperCase();
        return this.value.toLocaleString(locale, options);
    }
    toJSSource() {
        return `new __builtin.Currency(${this.value}, "${this.code}")`;
    }
}
exports.Currency = Currency;
/**
 * Runtime representation of a recurrent time rule (a single item in a recurrent time specification).
 *
 */
class RecurrentTimeRule {
    constructor({ beginTime, endTime, interval = 86400000, frequency = 1, dayOfWeek = null, beginDate = null, endDate = null, subtract = false }) {
        (0, assert_1.default)(beginTime instanceof Time);
        (0, assert_1.default)(endTime instanceof Time);
        (0, assert_1.default)(typeof interval === 'number');
        (0, assert_1.default)(typeof frequency === 'number');
        (0, assert_1.default)(dayOfWeek === null || (typeof dayOfWeek === 'number' && dayOfWeek >= 0 && dayOfWeek <= 6));
        (0, assert_1.default)(beginDate === null || beginDate instanceof Date);
        (0, assert_1.default)(endDate === null || endDate instanceof Date);
        (0, assert_1.default)(typeof subtract === 'boolean');
        this.beginTime = beginTime;
        this.endTime = endTime;
        this.interval = interval;
        this.frequency = frequency;
        this.dayOfWeek = dayOfWeek;
        this.beginDate = beginDate;
        this.endDate = endDate;
        this.subtract = subtract;
    }
    toString() {
        return `RecurrentTimeRule(${this.beginTime} -- ${this.endTime}; ${this.frequency} every ${this.interval} ms; from ${this.beginDate} to ${this.endDate})`;
    }
    toJSSource() {
        return `new __builtin.RecurrentTimeRule({ `
            + `beginTime: ${this.beginTime.toJSSource()}, `
            + `endTime: ${this.endTime.toJSSource()}, `
            + `interval: ${this.interval}, `
            + `frequency: ${this.frequency}, `
            + `dayOfWeek: ${this.dayOfWeek}, `
            + 'beginDate: ' + (this.beginDate ? `new Date(${this.beginDate.getTime()})` : 'null') + ', '
            + 'endDate: ' + (this.endDate ? `new Date(${this.endDate.getTime()})` : 'null') + ', '
            + `subtract: ${this.subtract}, `
            + '})';
    }
    contains(dateOrTime, timezone) {
        let time;
        let instant;
        let dayOfWeek;
        if (dateOrTime instanceof Date) {
            instant = polyfill_1.toTemporalInstant.call(dateOrTime);
            const datetz = instant.toZonedDateTimeISO(timezone);
            // this is not correct wrt timezone but if all we have is a date that's all we can do
            time = new Time(datetz.hour, datetz.minute, datetz.second);
            dayOfWeek = datetz.dayOfWeek;
        }
        else if (dateOrTime instanceof polyfill_1.Temporal.ZonedDateTime) {
            time = Time.fromTemporal(dateOrTime.toPlainTime());
            dayOfWeek = dateOrTime.dayOfWeek;
            instant = dateOrTime.toInstant();
        }
        else if (dateOrTime instanceof polyfill_1.Temporal.PlainTime) {
            time = Time.fromTemporal(dateOrTime);
            dayOfWeek = undefined;
            instant = undefined;
        }
        else {
            time = dateOrTime;
            dayOfWeek = undefined;
            instant = undefined;
        }
        if (!(this.beginTime <= time && this.endTime >= time))
            return false;
        if (instant) {
            if (this.beginDate && this.beginDate.getTime() > instant.epochMilliseconds)
                return false;
            if (this.endDate && this.endDate.getTime() < instant.epochMilliseconds)
                return false;
        }
        if (dayOfWeek) {
            if (this.dayOfWeek !== null && this.dayOfWeek !== dayOfWeek)
                return false;
        }
        // TODO frequency and interval
        return true;
    }
}
exports.RecurrentTimeRule = RecurrentTimeRule;
class Aggregation {
    constructor(type, field, cols, count) {
        this.type = type;
        this.field = field;
        this.cols = cols;
        this.count = count;
    }
}
exports.Aggregation = Aggregation;
