// genie-toolkit/dist/lib/dialogue-agent/handlers/3rdparty.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2021 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Tp = __importStar(require("thingpedia"));
const dialogue_loop_1 = require("../dialogue-loop");
const value_category_1 = __importDefault(require("../value-category"));
/**
 * Helper class to adapt a Thingpedia device into a Genie dialogue handler.
 *
 * Thingpedia dialogue handlers are simplified to avoid leaking some details
 * of the ThingTalk dialogue loop.
 */
class ThingpediaDialogueHandler {
    constructor(device) {
        this.uniqueId = device.uniqueId;
        this._iface = device.queryInterface('dialogue-handler');
    }
    get priority() {
        return this._iface.priority;
    }
    get icon() {
        return this._iface.icon;
    }
    _mapReplyResult(result) {
        return {
            expecting: result.expecting ? value_category_1.default.fromType(result.expecting) : null,
            messages: result.messages,
            context: result.context,
            agent_target: result.agent_target
        };
    }
    _mapConfidence(confidence) {
        switch (confidence) {
            case Tp.DialogueHandler.Confidence.EXACT_IN_DOMAIN_COMMAND:
                return dialogue_loop_1.CommandAnalysisType.EXACT_IN_DOMAIN_COMMAND;
            case Tp.DialogueHandler.Confidence.STRONGLY_CONFIDENT_IN_DOMAIN_COMMAND: // FIXME
                return dialogue_loop_1.CommandAnalysisType.STRONGLY_CONFIDENT_IN_DOMAIN_COMMAND;
            case Tp.DialogueHandler.Confidence.CONFIDENT_IN_DOMAIN_COMMAND:
                return dialogue_loop_1.CommandAnalysisType.CONFIDENT_IN_DOMAIN_COMMAND;
            case Tp.DialogueHandler.Confidence.EXACT_IN_DOMAIN_FOLLOWUP:
                return dialogue_loop_1.CommandAnalysisType.EXACT_IN_DOMAIN_FOLLOWUP;
            case Tp.DialogueHandler.Confidence.STRONGLY_CONFIDENT_IN_DOMAIN_FOLLOWUP:
                return dialogue_loop_1.CommandAnalysisType.STRONGLY_CONFIDENT_IN_DOMAIN_FOLLOWUP;
            case Tp.DialogueHandler.Confidence.CONFIDENT_IN_DOMAIN_FOLLOWUP:
                return dialogue_loop_1.CommandAnalysisType.CONFIDENT_IN_DOMAIN_FOLLOWUP;
            case Tp.DialogueHandler.Confidence.NONCONFIDENT_IN_DOMAIN_COMMAND:
                return dialogue_loop_1.CommandAnalysisType.NONCONFIDENT_IN_DOMAIN_COMMAND;
            case Tp.DialogueHandler.Confidence.NONCONFIDENT_IN_DOMAIN_FOLLOWUP:
                return dialogue_loop_1.CommandAnalysisType.NONCONFIDENT_IN_DOMAIN_FOLLOWUP;
            case Tp.DialogueHandler.Confidence.OUT_OF_DOMAIN_COMMAND:
                return dialogue_loop_1.CommandAnalysisType.OUT_OF_DOMAIN_COMMAND;
            default:
                throw new TypeError();
        }
    }
    async initialize(initialState, showWelcome) {
        const result = await this._iface.initialize(initialState, showWelcome);
        if (result === null)
            return null;
        return this._mapReplyResult(result);
    }
    getState() {
        return this._iface.getState();
    }
    reset() {
        return this._iface.reset();
    }
    async analyzeCommand(command) {
        if (command.type !== 'command') {
            return {
                type: dialogue_loop_1.CommandAnalysisType.NONCONFIDENT_IN_DOMAIN_COMMAND,
                utterance: '',
                user_target: '',
                inner: null
            };
        }
        const inner = await this._iface.analyzeCommand(command.utterance);
        return {
            type: this._mapConfidence(inner.confident),
            utterance: inner.utterance,
            user_target: inner.user_target,
            inner
        };
    }
    async getReply(command) {
        return this._mapReplyResult(await this._iface.getReply(command.inner));
    }
    async getFollowUp() {
        return null;
    }
}
exports.default = ThingpediaDialogueHandler;
