// thingtalk/dist/ast/dialogues.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of ThingTalk
//
// Copyright 2020 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialogueState = exports.DialogueHistoryItem = exports.DialogueHistoryResultList = exports.DialogueHistoryResultItem = void 0;
const assert_1 = __importDefault(require("assert"));
const base_1 = __importStar(require("./base"));
const program_1 = require("./program");
const statement_1 = require("./statement");
const Optimizer = __importStar(require("../optimize"));
const typecheck_1 = __importDefault(require("../typecheck"));
const values_1 = require("./values");
const visitor_1 = __importDefault(require("./visitor"));
const slots_1 = require("./slots");
const list_1 = __importDefault(require("../utils/list"));
class DialogueHistoryResultItem extends base_1.default {
    constructor(location, value, raw = null) {
        super(location);
        (0, assert_1.default)(typeof value === 'object');
        this.value = value;
        (0, assert_1.default)(raw === null || typeof raw === 'object');
        this.raw = raw;
    }
    toSource() {
        return new values_1.Value.Object(this.value).toSource();
    }
    clone() {
        const newValue = {};
        Object.assign(newValue, this.value);
        return new DialogueHistoryResultItem(this.location, newValue, this.raw);
    }
    visit(visitor) {
        visitor.enter(this);
        if (visitor.visitDialogueHistoryResultItem(this)) {
            for (const key in this.value)
                this.value[key].visit(visitor);
        }
        visitor.exit(this);
    }
    *iterateSlots2(schema) {
        for (const key in this.value) {
            const arg = (schema ? schema.getArgument(key) : null) || null;
            yield* (0, slots_1.recursiveYieldArraySlots)(new slots_1.ResultSlot(null, {}, arg, this.value, key));
        }
    }
    equals(other) {
        const keys = Object.keys(this.value).sort();
        const otherkeys = Object.keys(other.value).sort();
        if (keys.length !== otherkeys.length)
            return false;
        for (let i = 0; i < keys.length; i++) {
            if (keys[i] !== otherkeys[i])
                return false;
            if (!this.value[keys[i]].equals(other.value[otherkeys[i]]))
                return false;
        }
        return true;
    }
}
exports.DialogueHistoryResultItem = DialogueHistoryResultItem;
function setEquals(set1, set2) {
    if (set1.size !== set2.size)
        return false;
    for (const el of set1) {
        if (!set2.has(el))
            return false;
    }
    return true;
}
/**
 * A wrapper for results that contains the current page of results to present, the
 * total number of results and whether there are more results that have not been
 * fetched.
 *
 */
class DialogueHistoryResultList extends base_1.default {
    constructor(location, results, count, more = false, error = null) {
        super(location);
        (0, assert_1.default)(Array.isArray(results));
        (0, assert_1.default)(count instanceof values_1.Value);
        // either count is not a number (it's a __const_ token) or it's at least as many as the results we see
        (0, assert_1.default)(!(count instanceof values_1.NumberValue) || count.value >= results.length);
        // at least one results is always presented, unless there are truly no results
        (0, assert_1.default)(results.length > 0 || (count instanceof values_1.NumberValue && count.value === 0));
        (0, assert_1.default)(typeof more === 'boolean');
        this.results = results;
        this.count = count;
        this.more = more;
        this.error = error;
    }
    toSource() {
        let list;
        if (this.results.length === 0) {
            list = list_1.default.concat('#[', 'results', '=', '[', ']', ']');
        }
        else {
            list = list_1.default.concat('#[', 'results', '=', '[', '\n', '\t+');
            let first = true;
            for (const result of this.results) {
                if (first)
                    first = false;
                else
                    list = list_1.default.concat(list, ',', '\n');
                list = list_1.default.concat(list, result.toSource());
            }
            list = list_1.default.concat(list, '\n', '\t-', ']', ']');
        }
        if (!(this.count instanceof values_1.Value.Number && this.count.value <= this.results.length))
            list = list_1.default.concat(list, '\n', '#[', 'count', '=', this.count.toSource(), ']');
        if (this.more)
            list = list_1.default.concat(list, '\n', '#[', 'more', '=', 'true', ']');
        if (this.error)
            list = list_1.default.concat(list, '\n', '#[', 'error', '=', this.error.toSource(), ']');
        return list;
    }
    clone() {
        return new DialogueHistoryResultList(this.location, this.results.map((r) => r.clone()), this.count.clone(), this.more, this.error);
    }
    visit(visitor) {
        visitor.enter(this);
        if (visitor.visitDialogueHistoryResultList(this)) {
            for (const result of this.results)
                result.visit(visitor);
            if (this.error)
                this.error.visit(visitor);
        }
        visitor.exit(this);
    }
    *iterateSlots2(schema) {
        for (const result of this.results)
            yield* result.iterateSlots2(schema);
    }
    equals(other) {
        if (this === other)
            return true;
        if (this.more !== other.more || !this.count.equals(other.count))
            return false;
        if (this.error !== other.error) {
            if (!this.error || !other.error)
                return false;
            if (!this.error.equals(other.error))
                return false;
        }
        if (this.results.length !== other.results.length)
            return false;
        for (let i = 0; i < this.results.length; i++) {
            if (!this.results[i].equals(other.results[i]))
                return false;
        }
        return true;
    }
}
exports.DialogueHistoryResultList = DialogueHistoryResultList;
/**
 * A single item in the dialogue state. Consists of a program and optionally
 * the results from that program.
 *
 */
class DialogueHistoryItem extends base_1.default {
    constructor(location, stmt, results, confirm, { nl, impl } = {}) {
        super(location);
        (0, assert_1.default)(stmt instanceof statement_1.ExpressionStatement);
        (0, assert_1.default)(results === null || results instanceof DialogueHistoryResultList);
        if (typeof confirm === 'boolean')
            confirm = confirm ? 'confirmed' : 'accepted';
        (0, assert_1.default)(confirm === 'proposed' || confirm === 'accepted' || confirm === 'confirmed');
        (0, assert_1.default)(confirm === 'confirmed' || results === null);
        this.stmt = stmt;
        this.results = results;
        this.confirm = confirm;
        this.nl_annotations = nl || {};
        this.impl_annotations = impl || {};
    }
    toSource() {
        // ensure that recursive toSource() calls occur in source order,
        // so values are fetched from the entity allocator in the right order
        const expression = this.stmt.expression.toSource();
        const results = this.results !== null ? this.results.toSource() : null;
        const annotations = list_1.default.concat((0, base_1.nlAnnotationsToSource)(this.nl_annotations), (0, base_1.implAnnotationsToSource)(this.impl_annotations));
        // note: we punch through to stmt.expression because stmt.toSource() will
        // add the semicolon, which we don't want
        if (results !== null)
            return list_1.default.concat(expression, '\n', results, annotations, ';');
        else if (this.confirm !== 'accepted' || annotations !== list_1.default.Nil)
            return list_1.default.concat(expression, '\n', '#[', 'confirm', '=', new values_1.Value.Enum(this.confirm).toSource(), ']', annotations, ';');
        else
            return list_1.default.concat(expression, annotations, ';');
    }
    _getFunctions() {
        const functions = new Set;
        const visitor = new class extends visitor_1.default {
            visitInvocation(invocation) {
                functions.add(`call:${invocation.selector.kind}:${invocation.channel}`);
                return false;
            }
        };
        this.stmt.visit(visitor);
        return functions;
    }
    compatible(other) {
        return setEquals(this._getFunctions(), other._getFunctions());
    }
    isExecutable() {
        return this.stmt.isExecutable();
    }
    equals(other) {
        if (this === other)
            return true;
        if (this.confirm !== other.confirm)
            return false;
        // HACK prettyprint to compare for equality is quite expensive, we should open-code
        // equality properly
        if (this.stmt.prettyprint() !== other.stmt.prettyprint())
            return false;
        if (this.results === other.results)
            return true;
        if (this.results === null || other.results === null)
            return false;
        return this.results.equals(other.results);
    }
    optimize() {
        const newStmt = Optimizer.optimizeRule(this.stmt);
        this.stmt = newStmt;
        return this;
    }
    clone() {
        // clone annotations
        const nl = {};
        Object.assign(nl, this.nl_annotations);
        const impl = {};
        Object.assign(impl, this.impl_annotations);
        const annotations = { nl, impl };
        return new DialogueHistoryItem(this.location, this.stmt.clone(), this.results ? this.results.clone() : null, this.confirm, annotations);
    }
    visit(visitor) {
        visitor.enter(this);
        if (visitor.visitDialogueHistoryItem(this)) {
            this.stmt.visit(visitor);
            if (this.results !== null)
                this.results.visit(visitor);
        }
        visitor.exit(this);
    }
    *iterateSlots() {
        yield* this.stmt.iterateSlots();
        // no slots in a HistoryResult
    }
    *iterateSlots2() {
        yield* this.stmt.iterateSlots2();
        if (this.results === null)
            return;
        const stmtSchema = this.stmt.expression.schema;
        yield* this.results.iterateSlots2(stmtSchema);
    }
}
exports.DialogueHistoryItem = DialogueHistoryItem;
/**
 * The representation of a dialogue state.
 *
 * It is composed of:
 *
 * - A policy name (which is used as a namespace for the dialogueAct)
 * - The current abstract dialogue act
 * - An parameter name that modifies the current dialogue act (optional, used for questions from the agent)
 * - The history of all programs up to that point, and all programs that are scheduled to execute
 *
 */
class DialogueState extends program_1.Input {
    constructor(location, policy, dialogueAct, dialogueActParam, history, { nl, impl } = {}) {
        super(location);
        (0, assert_1.default)(typeof policy === 'string');
        (0, assert_1.default)(typeof dialogueAct === 'string');
        (0, assert_1.default)(dialogueActParam === null || typeof dialogueActParam === 'string' || Array.isArray(dialogueActParam));
        this.dialogueAct = dialogueAct;
        this.dialogueActParam = typeof dialogueActParam === 'string' ? [dialogueActParam] : dialogueActParam;
        this.history = history;
        this.policy = policy;
        this.nl_annotations = nl || {};
        this.impl_annotations = impl || {};
        this._current = null;
        this._next = null;
        this.updateCurrent();
    }
    /**
     * A pointer to the most recently executed ThingTalk statement, if any.
     *
     * This is a pointer into the history, provided for API convenience only.
     */
    get current() {
        return this._current;
    }
    /**
     * The definition of the main Thingpedia function associated with the current
     * statement, if any.
     */
    get currentFunction() {
        return this._current ? this._current.stmt.expression.schema : null;
    }
    /**
     * The definition of the main Thingpedia query associated with the current
     * statement, if any.
     *
     * If the current statement is a query, this will be the same as
     * {@link currentFunction}. Otherwise, it will be the last query before
     * the action is invoked.
     */
    get currentQuery() {
        if (!this._current)
            return null;
        const lastQuery = this._current.stmt.expression.lastQuery;
        if (!lastQuery)
            return null;
        return lastQuery.schema;
    }
    /**
     * The results of the current ThingTalk statements, if any.
     *
     * This is a convenience API to access the results of the current statement.
     */
    get currentResults() {
        return this._current ? this._current.results : null;
    }
    /**
     * A pointer to the immediate next unexecuted ThingTalk statement, if any.
     *
     * This is a pointer into the history, provided for API convenience only.
     */
    get next() {
        return this._next;
    }
    /**
     * The definition of the main Thingpedia function associated with the next
     * statement, if any.
     */
    get nextFunction() {
        return this._next ? this._next.stmt.expression.schema : null;
    }
    /**
     * Update the {@link current} and {@link next} pointers.
     *
     * This method should be called when {@link history} is modified or statements
     * are executed.
     */
    updateCurrent() {
        for (const item of this.history) {
            if (item.confirm === 'proposed')
                continue;
            if (item.results === null) {
                this._next = item;
                break;
            }
            this._current = item;
        }
    }
    optimize() {
        this.history = this.history.map((prog) => prog.optimize())
            .filter((prog) => prog !== null);
        return this;
    }
    toSource() {
        let list = list_1.default.concat('$dialogue', '@' + this.policy, '.', this.dialogueAct);
        if (this.dialogueActParam)
            list = list_1.default.concat(list, '(', list_1.default.join(this.dialogueActParam.map((p) => typeof p === 'string' ? list_1.default.singleton(p) : p.toSource()), ','), ')');
        const annotations = list_1.default.concat((0, base_1.nlAnnotationsToSource)(this.nl_annotations), (0, base_1.implAnnotationsToSource)(this.impl_annotations));
        list = list_1.default.concat(list, annotations, ';');
        for (const item of this.history)
            list = list_1.default.concat(list, '\n', item.toSource());
        return list;
    }
    clone() {
        // clone annotations
        const nl = {};
        Object.assign(nl, this.nl_annotations);
        const impl = {};
        Object.assign(impl, this.impl_annotations);
        const annotations = { nl, impl };
        return new DialogueState(this.location, this.policy, this.dialogueAct, this.dialogueActParam ? this.dialogueActParam.map((v) => typeof v === 'string' ? v : v.clone()) : null, this.history.map((item) => item.clone()), annotations);
    }
    async typecheck(schemas, getMeta = false) {
        const typeChecker = new typecheck_1.default(schemas, getMeta);
        await typeChecker.typeCheckDialogue(this);
        return this;
    }
    visit(visitor) {
        visitor.enter(this);
        if (visitor.visitDialogueState(this)) {
            for (const item of this.history)
                item.visit(visitor);
        }
        visitor.exit(this);
    }
    *iterateSlots() {
        for (const item of this.history)
            yield* item.iterateSlots();
    }
    *iterateSlots2() {
        for (const item of this.history)
            yield* item.iterateSlots2();
    }
}
exports.DialogueState = DialogueState;
DialogueState.prototype.isDialogueState = true;
program_1.Input.DialogueState = DialogueState;
