// genie-toolkit/dist/lib/dialogue-agent/audio/protocol.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2021 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestType = void 0;
// Protocol types associated with the audio control protocol
var RequestType;
(function (RequestType) {
    RequestType["CHECK_BACKEND"] = "check";
    RequestType["PREPARE"] = "prepare";
    RequestType["STOP"] = "stop";
    RequestType["PAUSE"] = "pause";
    RequestType["RESUME"] = "resume";
    RequestType["PLAY_URLS"] = "play-urls";
    RequestType["SET_VOLUME"] = "set-volume";
    RequestType["ADJUST_VOLUME"] = "adj-volume";
    RequestType["SET_MUTE"] = "set-mute";
    RequestType["SET_VOICE_INPUT"] = "set-voice-input";
    RequestType["SET_VOICE_OUTPUT"] = "set-voice-output";
})(RequestType = exports.RequestType || (exports.RequestType = {}));
