// thingpedia/dist/loaders/base_ondisk_js.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Thingpedia
//
// Copyright 2019-2020 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
const tmp = __importStar(require("tmp"));
const util = __importStar(require("util"));
const Module = __importStar(require("module"));
const Helpers = __importStar(require("../helpers"));
const I18n = __importStar(require("../i18n"));
const base_js_1 = __importDefault(require("./base_js"));
function resolve(mainModule) {
    if (process.platform !== 'win32' && !mainModule.startsWith('/'))
        throw new Error('Invalid relative module path');
    if (require.resolve)
        return require.resolve(mainModule);
    else
        return Module._resolveFilename(mainModule, module, false);
}
function clearRequireCache(mainModule) {
    try {
        const fileName = resolve(mainModule);
        console.log(mainModule + ' was cached as ' + fileName);
        delete require.cache[fileName];
        const prefix = path.dirname(fileName) + '/';
        for (const key in require.cache) {
            if (key.startsWith(prefix))
                delete require.cache[key];
        }
    }
    catch (e) {
        // do nothing
    }
}
/**
 * Base class of all loaders that retrieve JS code on-demand from a
 * Thingpedia server and cache it on disk.
 *
 * This differs from {@link BaseJavascriptLoader} because the latter covers
 * {@link BuiltinLoader} too.
 */
class BaseOnDiskJavascriptLoader extends base_js_1.default {
    constructor(kind, manifest, parents, loader) {
        super(kind, manifest, parents, loader);
        this._platform = loader.platform;
        this._cacheDir = loader.platform.getCacheDir() + '/device-classes';
        this._modulePath = null;
        // for compat with thingpedia devices that have not been updated recently
        if (!this._manifest.annotations.package_version)
            this._manifest.annotations.package_version = this._manifest.annotations.version;
    }
    get package_version() {
        return this._manifest.annotations.package_version.toJS();
    }
    clearCache() {
        this._loading = null;
        if (this._modulePath)
            clearRequireCache(this._modulePath);
    }
    async _loadJsModule() {
        const modulePath = this._modulePath;
        const version = JSON.parse(fs.readFileSync(modulePath + '/package.json').toString('utf8'))['thingpedia-version'];
        if (version !== undefined && version !== this.package_version) {
            console.log(`Cached module ${this.id} is out of date (found ${version}, want ${this.package_version})`);
            return null;
        }
        let deviceClass = await Promise.resolve().then(() => __importStar(require(modulePath)));
        // ES Module compatibility for "export default"
        if (typeof deviceClass !== 'function' && typeof deviceClass.default === 'function')
            deviceClass = deviceClass.default;
        deviceClass.require = function (subpath) {
            return require(path.resolve(modulePath, subpath));
        };
        const modir = path.resolve(modulePath, 'po');
        const gettext = this._platform.getCapability('gettext');
        if (await util.promisify(fs.exists)(modir)) {
            if (gettext && this._platform.locale !== 'en-US')
                await I18n.loadTextdomainDirectory(gettext, this._platform.locale, this.id, modir);
        }
        if (gettext) {
            deviceClass.gettext = {
                gettext: gettext.dgettext.bind(gettext, this.id),
                ngettext: gettext.dngettext.bind(gettext, this.id)
            };
        }
        else {
            deviceClass.gettext = {
                gettext(x) {
                    return x;
                },
                ngettext(x1, xn, n) {
                    return n === 1 ? x1 : xn;
                },
            };
        }
        return this._completeLoading(deviceClass);
    }
    async _doGetDeviceClass() {
        this._modulePath = path.resolve(process.cwd(), this._cacheDir + '/' + this._id);
        try {
            if (fs.existsSync(this._modulePath)) {
                const deviceClass = await this._loadJsModule();
                if (deviceClass !== null)
                    return deviceClass;
            }
            if (!this._platform.hasCapability('code-download'))
                throw new Error('Code download is not allowed on this platform');
            const redirect = await this._client.getModuleLocation(this._id);
            if (redirect.startsWith('file:///') && !redirect.endsWith('.zip')) {
                this._modulePath = redirect.substring('file://'.length);
                return (await this._loadJsModule());
            }
            const response = await Helpers.Content.getStream(this._platform, redirect);
            const [path, fd] = await new Promise((resolve, reject) => {
                tmp.file({ mode: 0o600,
                    keep: true,
                    dir: this._platform.getTmpDir(),
                    prefix: 'thingengine-' + this._id + '-',
                    postfix: '.zip' }, (err, path, fd) => {
                    if (err)
                        reject(err);
                    else
                        resolve([path, fd]);
                });
            });
            const stream = fs.createWriteStream('', { fd, flags: 'w' });
            const zipPath = await new Promise((callback, errback) => {
                response.pipe(stream);
                stream.on('finish', () => {
                    callback(path);
                });
                stream.on('error', errback);
            });
            try {
                fs.mkdirSync(this._modulePath);
            }
            catch (e) {
                if (e.code !== 'EEXIST')
                    throw e;
            }
            const unzip = this._platform.getCapability('code-download');
            await unzip.unzip(zipPath, this._modulePath);
            fs.unlinkSync(zipPath);
            return (await this._loadJsModule());
        }
        catch (e) {
            this._loading = null;
            throw e;
        }
    }
}
exports.default = BaseOnDiskJavascriptLoader;
