// thingtalk/dist/new-syntax/nn-lexer.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of ThingTalk
//
// Copyright 2019-2020 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.nnLexer = void 0;
const assert = __importStar(require("assert"));
const errors_1 = require("../utils/errors");
const keywords_1 = require("./keywords");
const token_1 = require("./token");
const DECIMAL_LITERAL = /^-?(?:(?:0|[1-9][0-9]*)\.[0-9]*(?:[eE][+-]?[0-9]+)?|\.[0-9]+(?:[eE][+-]?[0-9]+)?|(?:0|[1-9][0-9]*)(?:[eE][+-]?[0-9]+)?)$/;
// matches one or more identifiers separated by a period
const IDENTIFIER = /^[A-Za-z_][A-Za-z0-9_]*(?:\.[A-Za-z_][A-Za-z0-9]*)*$/;
const OLD_ENTITY_REFERENCE = /Entity\s?\(\s?[A-Za-z_][A-Za-z0-9_-]*(?:\.[A-Za-z_][A-Za-z0-9_-]*)*:[A-Za-z_][A-Za-z0-9_]*\s?\)/;
function isEntity(token) {
    return /^(?:QUOTED_STRING|NUMBER|CURRENCY|DURATION|MEASURE|TIME|DATE|LOCATION|USERNAME|HASHTAG|URL|PHONE_NUMBER|EMAIL_ADDRESS|PATH_NAME|PICTURE|GENERIC_ENTITY|SLOT)_/.test(token);
}
function isIdentifierLike(token) {
    return IDENTIFIER.test(token);
}
function isUnit(token) {
    return (isIdentifierLike(token) && !keywords_1.KEYWORDS.has(token) && !keywords_1.CONTEXTUAL_KEYWORDS.has(token)) || token === 'in' || token === 'min';
}
class DummyLanguagePack {
    detokenizeSentence(words) {
        return words.join(' ');
    }
}
function* nnLexer(input, entities, languagePack = new DummyLanguagePack()) {
    let sequence;
    if (typeof input === 'string')
        sequence = input.split(' ');
    else
        sequence = input;
    let getEntity;
    if (typeof entities !== 'function') {
        getEntity = (next) => {
            if (!(next in entities)) {
                if (next.startsWith('SLOT_'))
                    return undefined;
                throw new errors_1.ThingTalkSyntaxError(`Invalid entity ${next}, have [${Object.keys(entities)}]`, makeLocation());
            }
            return entities[next];
        };
    }
    else {
        getEntity = entities;
    }
    let i = 0;
    let charoffset = 0;
    // FIXME
    const lastfunction = null;
    const lastparam = null;
    function makeLocation() {
        const tok = sequence[i];
        const start = {
            offset: 1 + charoffset,
            column: 1 + charoffset,
            line: 1,
            token: i
        };
        const end = {
            offset: 1 + charoffset + tok.length,
            column: 1 + charoffset + tok.length,
            line: 1,
            token: i + 1
        };
        return { start, end };
    }
    function advance() {
        const tok = sequence[i];
        if (i > 0)
            charoffset += 1;
        charoffset += tok.length;
        i++;
    }
    for (; i < sequence.length; advance()) {
        const next = sequence[i];
        if (next === '"') {
            const start = {
                offset: 1 + charoffset,
                column: 1 + charoffset,
                line: 1,
                token: i
            };
            const words = [];
            for (advance(); i < sequence.length; advance()) {
                if (sequence[i] === '"')
                    break;
                words.push(sequence[i]);
            }
            if (i >= sequence.length)
                throw new errors_1.ThingTalkSyntaxError(`Unterminated string literal`, makeLocation());
            assert.strictEqual(sequence[i], '"');
            const end = {
                offset: 1 + charoffset + '"'.length,
                column: 1 + charoffset + '"'.length,
                line: 1,
                token: i + 1
            };
            yield token_1.Token.make('QUOTED_STRING', { start, end }, languagePack.detokenizeSentence(words));
        }
        else if (next === '""') {
            yield token_1.Token.make('QUOTED_STRING', makeLocation(), '');
        }
        else if (keywords_1.FORBIDDEN_KEYWORDS.has(next)) {
            throw new errors_1.ThingTalkSyntaxError(`Use of forbidden token ${next}`, makeLocation());
        }
        else if (DECIMAL_LITERAL.test(next)) {
            yield token_1.Token.make('NUMBER', makeLocation(), parseFloat(next));
        }
        else if (isEntity(next)) {
            // check if we have a unit next, to pass to the entity retriever
            let unit = null;
            if (i < sequence.length - 1 && isUnit(sequence[i + 1]))
                unit = sequence[i + 1];
            // entity
            const entity = getEntity(next, lastparam, lastfunction, unit);
            const entityType = next.substring(0, next.lastIndexOf('_'));
            if (entityType.startsWith('GENERIC_ENTITY_')) {
                const generic = entity;
                yield token_1.Token.make('GENERIC_ENTITY', makeLocation(), {
                    value: generic.value,
                    display: generic.display,
                    type: entityType.substring('GENERIC_ENTITY_'.length)
                });
            }
            else if (entityType.startsWith('MEASURE_')) {
                yield token_1.Token.make('MEASURE', makeLocation(), entity);
            }
            else {
                yield token_1.Token.make(entityType, makeLocation(), entity);
            }
        }
        else if (next.startsWith('@')) {
            yield token_1.Token.make('CLASS_OR_FUNCTION_REF', makeLocation(), next.substring(1));
        }
        else if (next.startsWith('$')) {
            if (keywords_1.DOLLAR_KEYWORDS.has(next))
                yield token_1.Token.make(next, makeLocation(), null);
            else
                yield token_1.Token.make('DOLLARIDENTIFIER', makeLocation(), next.substring(1));
        }
        else if (next.startsWith('^^')) {
            yield token_1.Token.make('ENTITY_NAME', makeLocation(), next.substring(2));
        }
        else if (next === 'Entity') {
            yield token_1.Token.make(next, makeLocation(), null);
            if (i < sequence.length - 3 && OLD_ENTITY_REFERENCE.test(sequence.slice(i, i + 4).join(''))) {
                advance();
                yield token_1.Token.make('(', makeLocation(), null);
                advance();
                yield token_1.Token.make('ENTITY_NAME', makeLocation(), sequence[i]);
            }
        }
        else if (isIdentifierLike(next)) {
            const parts = next.split('.');
            let first = true;
            for (const part of parts) {
                if (first)
                    first = false;
                else
                    yield token_1.Token.make('.', makeLocation(), null);
                if (keywords_1.KEYWORDS.has(part) || keywords_1.CONTEXTUAL_KEYWORDS.has(part))
                    yield token_1.Token.make(part, makeLocation(), null);
                else
                    yield token_1.Token.make('IDENTIFIER', makeLocation(), part);
            }
        }
        else {
            yield token_1.Token.make(next, makeLocation(), null);
        }
    }
}
exports.nnLexer = nnLexer;
