// genie-toolkit/dist/tool/autoqa/lib/utils.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2020-2021 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Silei Xu <silei@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_ENTITIES = exports.titleCase = exports.snakecase = exports.camelcase = exports.cleanEnumValue = exports.loadClassDefs = exports.loadClassDef = void 0;
const assert_1 = __importDefault(require("assert"));
const util_1 = __importDefault(require("util"));
const fs = __importStar(require("fs"));
const ThingTalk = __importStar(require("thingtalk"));
async function loadClassDef(thingpedia, options = { locale: 'en-US', timezone: undefined }) {
    const classes = await loadClassDefs(thingpedia, options);
    (0, assert_1.default)(classes.length === 1);
    return classes[0];
}
exports.loadClassDef = loadClassDef;
async function loadClassDefs(thingpedia, options = { locale: 'en-US', timezone: undefined }) {
    const library = ThingTalk.Syntax.parse(await util_1.default.promisify(fs.readFile)(thingpedia, { encoding: 'utf8' }), ThingTalk.Syntax.SyntaxType.Normal, options);
    (0, assert_1.default)(library instanceof ThingTalk.Ast.Library);
    return library.classes;
}
exports.loadClassDefs = loadClassDefs;
function cleanEnumValue(v) {
    // replace dash with space
    v = v.replace(/-/g, ' ');
    // camelcase the value
    v = camelcase(v);
    // add underscore prefix if value starts with number
    if (/^\d.*/.test(v))
        v = '_' + v;
    // normalize accent
    v = v.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
    return v;
}
exports.cleanEnumValue = cleanEnumValue;
function camelcase(v) {
    return v.replace(/(?:^|\s+|-)[A-Za-z]/g, (letter) => letter.trim().toUpperCase());
}
exports.camelcase = camelcase;
function snakecase(v) {
    return v.replace(/[() _-]+/g, '_').toLowerCase();
}
exports.snakecase = snakecase;
function titleCase(v) {
    return v.split(' ').map((word) => word[0].toUpperCase() + word.substring(1)).join(' ');
}
exports.titleCase = titleCase;
const DEFAULT_ENTITIES = [
    { "type": "tt:contact", "name": "Contact Identity", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:contact_name", "name": "Contact Name", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:device", "name": "Device Name", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:email_address", "name": "Email Address", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:flow_token", "name": "Flow Identifier", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:function", "name": "Function Name", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:hashtag", "name": "Hashtag", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:path_name", "name": "Unix Path", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:phone_number", "name": "Phone Number", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:picture", "name": "Picture", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:program", "name": "Program", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:url", "name": "URL", "is_well_known": 1, "has_ner_support": 0 },
    { "type": "tt:username", "name": "Username", "is_well_known": 1, "has_ner_support": 0 }
];
exports.DEFAULT_ENTITIES = DEFAULT_ENTITIES;
