// thingtalk/dist/index.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of ThingTalk
//
// Copyright 2016-2020 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.List = exports.Operators = exports.Units = exports.Helper = exports.PermissionChecker = exports.Time = exports.Entity = exports.Location = exports.Builtin = exports.ExecEnvironment = exports.Runtime = exports.Compiler = exports.Syntax = exports.SchemaRetriever = exports.Type = exports.Ast = exports.version = void 0;
const Units = __importStar(require("thingtalk-units"));
exports.Units = Units;
const Ast = __importStar(require("./ast"));
exports.Ast = Ast;
const compiler_1 = __importDefault(require("./compiler"));
exports.Compiler = compiler_1.default;
__exportStar(require("./compiler"), exports);
const Syntax = __importStar(require("./syntax_api"));
exports.Syntax = Syntax;
const exec_environment_1 = require("./runtime/exec_environment");
Object.defineProperty(exports, "ExecEnvironment", { enumerable: true, get: function () { return exec_environment_1.ExecEnvironment; } });
const Runtime = __importStar(require("./runtime/exec_environment"));
exports.Runtime = Runtime;
const type_1 = __importDefault(require("./type"));
exports.Type = type_1.default;
const schema_1 = __importDefault(require("./schema"));
exports.SchemaRetriever = schema_1.default;
__exportStar(require("./schema"), exports);
const permission_checker_1 = __importDefault(require("./permission_checker"));
exports.PermissionChecker = permission_checker_1.default;
const Helper = __importStar(require("./helper"));
exports.Helper = Helper;
const Builtin = __importStar(require("./runtime/builtins"));
exports.Builtin = Builtin;
const values_1 = require("./runtime/values");
Object.defineProperty(exports, "Location", { enumerable: true, get: function () { return values_1.Location; } });
Object.defineProperty(exports, "Entity", { enumerable: true, get: function () { return values_1.Entity; } });
Object.defineProperty(exports, "Time", { enumerable: true, get: function () { return values_1.Time; } });
const Operators = __importStar(require("./operators"));
exports.Operators = Operators;
const list_1 = __importDefault(require("./utils/list"));
exports.List = list_1.default;
/**
 * Version information
 *
 */
const version = '2.1.0';
exports.version = version;
