// genie-toolkit/dist/lib/engine/db/dbproxy/index.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2021 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatabaseProxy = void 0;
const local_table_1 = __importDefault(require("./local_table"));
const sync_table_1 = __importDefault(require("./sync_table"));
const FIELD_NAMES = {
    app: ['code', 'state', 'name', 'description'],
    device: ['state'],
    channel: ['value']
};
class DatabaseProxy {
    constructor(config) {
        this.config = config;
        this._config = config;
    }
    ensureSchema() {
        // database initialization is done in the cloud
        return Promise.resolve();
    }
    getLocalTable(name) {
        return new local_table_1.default(name, this._config.baseUrl, this._config.accessToken);
    }
    getSyncTable(name) {
        return new sync_table_1.default(name, this._config.baseUrl, this._config.accessToken, FIELD_NAMES[name]);
    }
}
exports.DatabaseProxy = DatabaseProxy;
