// genie-toolkit/dist/tool/download-entity-values.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2019-2021 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.initArgparse = void 0;
const ThingTalk = __importStar(require("thingtalk"));
const fs = __importStar(require("fs"));
const Tp = __importStar(require("thingpedia"));
const path = __importStar(require("path"));
const util = __importStar(require("util"));
const StreamUtils = __importStar(require("../lib/utils/stream-utils"));
const progress_bar_1 = __importDefault(require("./lib/progress_bar"));
const argutils_1 = require("./lib/argutils");
function initArgparse(subparsers) {
    const parser = subparsers.add_parser('download-entity-values', {
        add_help: true,
        description: "Download one or more entity dataset from Thingpedia."
    });
    parser.add_argument('-l', '--locale', {
        required: false,
        default: 'en-US',
        help: `BGP 47 locale tag of the natural language to download the snapshot for (defaults to 'en-US', American English)`
    });
    parser.add_argument('--thingpedia-url', {
        required: false,
        help: `base URL of Thingpedia server to contact; defaults to '${argutils_1.DEFAULT_THINGPEDIA_URL}'`
    });
    parser.add_argument('--developer-key', {
        required: false,
        default: '',
        help: `developer key to use when contacting Thingpedia`
    });
    parser.add_argument('-d', '--output-dir', {
        required: true,
    });
    parser.add_argument('--type', {
        required: false,
        action: 'append',
        default: [],
        help: `identifier of the Thingpedia entity datasets to download (if omitted, all entities are downloaded)`
    });
    parser.add_argument('--manifest', {
        required: false,
        help: `write a parameter dataset manifest to this location (suitable for Genie)`
    });
    parser.add_argument('--append-manifest', {
        required: false,
        action: 'store_true',
        help: `append to the manifest instead of replacing`
    });
    parser.add_argument('--debug', {
        action: 'store_true',
        help: 'Enable debugging.',
        default: false
    });
    parser.add_argument('--no-debug', {
        action: 'store_false',
        dest: 'debug',
        help: 'Disable debugging.',
    });
}
exports.initArgparse = initArgparse;
async function execute(args) {
    if (!args.thingpedia_url)
        args.thingpedia_url = await (0, argutils_1.getConfig)('thingpedia.url', process.env.THINGPEDIA_URL || argutils_1.DEFAULT_THINGPEDIA_URL);
    if (!args.developer_key)
        args.developer_key = await (0, argutils_1.getConfig)('thingpedia.developer-key', process.env.THINGPEDIA_DEVELOPER_KEY || null);
    await util.promisify(fs.mkdir)(args.output_dir, { recursive: true });
    if (args.type.length === 0) {
        let listUrl = `${args.thingpedia_url}/api/v3/entities/all?locale=${args.locale}&thingtalk_version=${ThingTalk.version}`;
        if (args.developer_key)
            listUrl += '&developer_key=' + args.developer_key;
        args.type = JSON.parse(await Tp.Helpers.Http.get(listUrl, { accept: 'application/json' }))
            .data.filter((e) => !e.is_well_known && e.has_ner_support).map((e) => e.type);
    }
    let progbar;
    if (!args.debug) {
        progbar = new progress_bar_1.default(args.type.length);
        // issue an update now to show the progress bar
        progbar.update(0);
    }
    let manifest;
    if (args.manifest)
        manifest = fs.createWriteStream(args.manifest, { flags: args.append_manifest ? 'a' : 'w' });
    let progress = 0;
    for (const type of args.type) {
        let url = `${args.thingpedia_url}/api/v3/entities/list/${type}?locale=${args.locale}&thingtalk_version=${ThingTalk.version}`;
        if (args.developer_key)
            url += '&developer_key=' + args.developer_key;
        const outputpath = path.resolve(args.output_dir, type + '.json');
        const output = fs.createWriteStream(outputpath);
        if (manifest)
            manifest.write(`entity\t${args.locale}\t${type}\t${path.relative(path.dirname(args.manifest), outputpath)}\n`);
        (await Tp.Helpers.Http.getStream(url)).pipe(output);
        await StreamUtils.waitFinish(output);
        progress++;
        if (progbar)
            progbar.update(progress);
    }
    if (manifest) {
        manifest.end();
        await StreamUtils.waitFinish(manifest);
    }
}
exports.execute = execute;
