// genie-toolkit/dist/tool/extract-translatable-annotations.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2021 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
//         Mehrad Moradshahi <mehrad@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.initArgparse = void 0;
const assert_1 = __importDefault(require("assert"));
const fs = __importStar(require("fs"));
const fs_1 = require("fs");
const ThingTalk = __importStar(require("thingtalk"));
const escaping_1 = require("../lib/utils/escaping");
const extract_translatable_1 = require("./lib/extract-translatable");
function initArgparse(subparsers) {
    const parser = subparsers.add_parser('extract-translatable-annotations', {
        add_help: true,
        description: "Extract translatable annotations from a thingpedia file into a JS file to be processed by xgettext."
    });
    parser.add_argument('input_file', {
        help: 'Input Thingpedia file to read from'
    });
    parser.add_argument('-o', '--output', {
        required: true,
        help: 'Output file to write to'
    });
}
exports.initArgparse = initArgparse;
async function execute(args) {
    const code = (await fs_1.promises.readFile(args.input_file)).toString();
    const parsed = ThingTalk.Syntax.parse(code, ThingTalk.Syntax.SyntaxType.Normal, {
        locale: 'en-US',
        timezone: 'UTC'
    });
    (0, assert_1.default)(parsed instanceof ThingTalk.Ast.Library);
    const output = fs.createWriteStream(args.output);
    for (const string of (0, extract_translatable_1.processLibrary)(parsed)) {
        output.write(`/* ${string.key} */\n`);
        if (string.comment)
            output.write(`/* ${string.comment} */\n`);
        if (string.context)
            output.write(`let x = pgettext(${(0, escaping_1.stringEscape)(string.context)}, ${(0, escaping_1.stringEscape)(string.object[string.field])});\n`);
        else
            output.write(`let x = _(${(0, escaping_1.stringEscape)(string.object[string.field])});\n`);
    }
    output.end();
}
exports.execute = execute;
