// genie-toolkit/dist/tool/init-device.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2019-2021 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.initArgparse = void 0;
const fs_1 = require("fs");
const path = __importStar(require("path"));
const process_utils_1 = require("../lib/utils/process-utils");
function initArgparse(subparsers) {
    const parser = subparsers.add_parser('init-device', {
        add_help: true,
        description: "Initialize the skeleton for a new Thingpedia device."
    });
    parser.add_argument('--loader', {
        required: false,
        default: 'org.thingpedia.v2',
        help: "The loader to use for the new device (defaults to org.thingpedia.v2)"
    });
    parser.add_argument('--description', {
        help: "The human-readable description of the new device",
        default: 'TODO add description'
    });
    parser.add_argument('--name', {
        help: "The human-readable name of the new device",
    });
    parser.add_argument('class_name', {
        help: "The name (unique ID) of the new device, in reverse-DNS notation (e.g. com.example.foo)"
    });
}
exports.initArgparse = initArgparse;
async function execute(args) {
    try {
        await fs_1.promises.rmdir(args.class_name);
    }
    catch (e) {
        if (e.code !== 'ENOENT') {
            console.error(`${args.class_name} already exists and is not an empty directory`);
            return;
        }
    }
    const packageInfo = JSON.parse(await fs_1.promises.readFile('./package.json', { encoding: 'utf8' }));
    await fs_1.promises.mkdir(args.class_name);
    await fs_1.promises.writeFile(path.resolve(args.class_name, 'manifest.tt'), `class @${args.class_name}
#_[name="${args.name || args.class_name}"]
#_[description="${args.description}"]
#_[thingpedia_name="${args.name || args.class_name}"]
#_[thingpedia_description="${args.description}"]
#[license="${packageInfo.license}"]
#[license_gplcompatible=${packageInfo.license !== 'Proprietary'}]
#[subcategory="service"] {
import loader from @${args.loader}();
import config from @org.thingpedia.config.none();
}
`);
    await fs_1.promises.writeFile(path.resolve(args.class_name, 'dataset.tt'), `dataset @${args.class_name} {
}
`);
    await fs_1.promises.mkdir(path.resolve(args.class_name, 'eval'));
    await fs_1.promises.mkdir(path.resolve(args.class_name, 'eval/dev'));
    await fs_1.promises.writeFile(path.resolve(args.class_name, 'eval/dev/annotated.txt'), '');
    await fs_1.promises.mkdir(path.resolve(args.class_name, 'eval/train'));
    await fs_1.promises.writeFile(path.resolve(args.class_name, 'eval/train/annotated.txt'), '');
    await fs_1.promises.writeFile(path.resolve(args.class_name, 'eval/paraphrase.tsv'), '');
    if (args.loader === 'org.thingpedia.v2') {
        await fs_1.promises.writeFile(path.resolve(args.class_name, 'package.json'), JSON.stringify({
            name: args.class_name,
            description: args.name || '',
            author: packageInfo.author,
            license: packageInfo.license,
            main: 'index.js'
        }));
        await fs_1.promises.writeFile(path.resolve(args.class_name, 'index.js'), `// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of ${args.class_name}
//
// Copyright ${(new Date).getFullYear()} ${packageInfo.author}
//
// See LICENSE for details
"use strict";

const Tp = require('thingpedia');

module.exports = class extends Tp.BaseDevice {
};
`);
        await (0, process_utils_1.execCommand)(['npm', 'install'], { debug: true, cwd: args.class_name });
    }
    await (0, process_utils_1.execCommand)(['git', 'add', args.class_name], { debug: true, });
    await fs_1.promises.writeFile(`test/unit/${args.class_name}.js`, `// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// Copyright ${(new Date).getFullYear()} ${packageInfo.author}
//
// See LICENSE for details
"use strict";

module.exports = [
];
`);
    await (0, process_utils_1.execCommand)(['git', 'add', `test/unit/${args.class_name}.js`], { debug: true, });
    await (0, process_utils_1.execCommand)(['git', 'commit', '-m', `Added ${args.class_name}`], { debug: true, });
    console.log('Success!');
}
exports.execute = execute;
