// genie-toolkit/dist/tool/translate-schema-annotations.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2021 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Mehrad Moradshahi <mehrad@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.initArgparse = void 0;
const node_gettext_1 = __importDefault(require("node-gettext"));
const gettextParser = __importStar(require("gettext-parser"));
const assert_1 = __importDefault(require("assert"));
const fs = __importStar(require("fs"));
const fs_1 = require("fs");
const ThingTalk = __importStar(require("thingtalk"));
const StreamUtils = __importStar(require("../lib/utils/stream-utils"));
const extract_translatable_1 = require("./lib/extract-translatable");
function initArgparse(subparsers) {
    const parser = subparsers.add_parser('translate-schema-annotations', {
        add_help: true,
        description: "Translate a thingpedia file's annotations to another language."
    });
    parser.add_argument('--po-file', {
        required: true,
        help: 'PO file containing original and translated strings'
    });
    parser.add_argument('input_file', {
        help: 'Input Thingpedia file to read'
    });
    parser.add_argument('-o', '--output', {
        required: true,
        type: fs.createWriteStream
    });
}
exports.initArgparse = initArgparse;
async function execute(args) {
    const code = (await fs_1.promises.readFile(args.input_file)).toString();
    const parsed = ThingTalk.Syntax.parse(code, ThingTalk.Syntax.SyntaxType.Normal, {
        locale: 'en-US',
        timezone: 'UTC'
    });
    (0, assert_1.default)(parsed instanceof ThingTalk.Ast.Library);
    const gt = new node_gettext_1.default();
    gt.textdomain('schema');
    try {
        const pofile = gettextParser.po.parse(await fs_1.promises.readFile(args.po_file, { encoding: 'utf8' }));
        gt.setLocale(pofile.headers.Language);
        gt.addTranslations(pofile.headers.Language, 'schema', pofile);
    }
    catch (e) {
        if (e.code !== 'ENOENT')
            throw e;
        // ignore if the file does not exist
        // this allows using this script for languages that are not supported yet
        // e.g. in a repository where some skills are translated and some are not
        console.log(`WARNING: ${args.po_file} does not exist, no translation will be applied`);
    }
    for (const entry of (0, extract_translatable_1.processLibrary)(parsed)) {
        if (entry.context)
            entry.object[entry.field] = gt.pgettext(entry.context, entry.object[entry.field]);
        else
            entry.object[entry.field] = gt.gettext(entry.object[entry.field]);
    }
    if (args.output) {
        args.output.end(parsed.prettyprint());
        await StreamUtils.waitFinish(args.output);
    }
}
exports.execute = execute;
