// thingpedia/dist/messaging.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Thingpedia
//
// Copyright 2016-2019 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const events = __importStar(require("events"));
const ref_counted_1 = __importDefault(require("./helpers/ref_counted"));
class Messaging extends events.EventEmitter {
    constructor() {
        super();
    }
    get isAvailable() {
        return false;
    }
    /* istanbul ignore next */
    get account() {
        throw new Error('Not Implemented');
    }
    /* istanbul ignore next */
    getIdentities() {
        throw new Error('Not Implemented');
    }
    /* istanbul ignore next */
    getFeedList() {
        throw new Error('Not Implemented');
    }
    /* istanbul ignore next */
    getFeed(feedId) {
        throw new Error('Not Implemented');
    }
    /* istanbul ignore next */
    getFeedWithContact(contactId) {
        throw new Error('Not Implemented');
    }
    /* istanbul ignore next */
    searchAccountByName(name) {
        throw new Error('Not Implemented');
    }
    /* istanbul ignore next */
    getAccountForIdentity(identity) {
        throw new Error('Not Implemented');
    }
}
(function (Messaging) {
    class Feed extends ref_counted_1.default {
        constructor(feedId) {
            super();
            this.feedId = feedId;
        }
        /* istanbul ignore next */
        async _doOpen() {
            throw new Error('Not Implemented');
        }
        /* istanbul ignore next */
        async _doClose() {
            throw new Error('Not Implemented');
        }
        /* istanbul ignore next */
        async sendText() {
            throw new Error('Not Implemented');
        }
        /* istanbul ignore next */
        async sendItem() {
            throw new Error('Not Implemented');
        }
        /* istanbul ignore next */
        async sendRaw() {
            throw new Error('Not Implemented');
        }
    }
    Messaging.Feed = Feed;
})(Messaging || (Messaging = {}));
exports.default = Messaging;
