// genie-toolkit/dist/tool/workers/augment-worker.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2019-2020 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ThingTalk = __importStar(require("thingtalk"));
const seedrandom_1 = __importDefault(require("seedrandom"));
const Tp = __importStar(require("thingpedia"));
const augmentation_1 = __importDefault(require("../../lib/dataset-tools/augmentation"));
async function worker(args, shard) {
    const tpClient = new Tp.FileClient(args);
    const schemaRetriever = new ThingTalk.SchemaRetriever(tpClient, null, args.debug);
    const constProvider = new Tp.FileParameterProvider(args.parameter_datasets, args.param_locale);
    await constProvider.load();
    return new augmentation_1.default(schemaRetriever, constProvider, tpClient, {
        rng: seedrandom_1.default.alea(args.random_seed + ':' + shard),
        locale: args.locale,
        timezone: args.timezone,
        paramLocale: args.param_locale,
        debug: args.debug,
        includeQuotedExample: false,
        quotedProbability: args.quoted_fraction,
        untypedStringProbability: args.untyped_string_probability,
        maxSpanLength: args.max_span_length,
        syntheticExpandFactor: args.synthetic_expand_factor,
        paraphrasingExpandFactor: args.quoted_paraphrasing_expand_factor,
        noQuoteExpandFactor: args.no_quote_paraphrasing_expand_factor,
        singleDeviceExpandFactor: args.single_device_expand_factor,
        cleanParameters: args.clean_parameters,
        requotable: args.requotable,
        includeEntityValue: args.include_entity_value,
        excludeEntityDisplay: args.exclude_entity_display,
        samplingType: args.sampling_type,
        subsetParamSet: args.subset_param_set.split('-'),
        numAttempts: args.num_attempts
    });
}
exports.default = worker;
