// string-interp/dist/ast.js
"use strict";
// -*- mode: ts; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// Copyright 2019-2020 The Board of Trustees of the Leland Stanford Junior University
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
//
// See LICENSE for details
Object.defineProperty(exports, "__esModule", { value: true });
exports.Select = exports.Plural = exports.Optional = exports.Placeholder = void 0;
class Placeholder {
    constructor(param, arraystyle, option = '', precision = 2, _default = null) {
        this.param = param;
        this.arraystyle = arraystyle;
        this.option = option;
        this.precision = precision;
        this.default = _default;
    }
}
exports.Placeholder = Placeholder;
class Optional {
    constructor(pattern) {
        this.pattern = pattern;
    }
}
exports.Optional = Optional;
class Plural {
    constructor(param, type, variants) {
        this.param = param;
        this.type = type;
        this.variants = variants;
    }
}
exports.Plural = Plural;
class Select {
    constructor(param, variants) {
        this.param = param;
        this.variants = variants;
    }
}
exports.Select = Select;
