// thingpedia/dist/loaders/unsupported_builtin.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Thingpedia
//
// Copyright 2019-2020 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ThingTalk = __importStar(require("thingtalk"));
const compat_1 = require("../compat");
const errors_1 = require("../errors");
const base_device_1 = __importDefault(require("../base_device"));
const base_1 = __importDefault(require("./base"));
/**
 * Loader for device classes that should be builtin but are
 * not available.
 */
class UnsupportedBuiltinModule extends base_1.default {
    constructor(kind, manifest, parents, loader) {
        // version does not matter for builtin modules
        manifest.annotations.version = new ThingTalk.Ast.Value.Number(0);
        super(kind, manifest, parents);
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        this._loaded = class UnsupportedDevice extends base_device_1.default {
            constructor(engine, state) {
                super(engine, state);
                this.uniqueId = this.kind;
            }
            async checkAvailable() {
                return base_device_1.default.Availability.OWNER_UNAVAILABLE;
            }
        };
        for (const [action,] of this._iterateFunctions(this._manifest, 'actions')) {
            this._loaded.prototype['do_' + action] = function () {
                throw new errors_1.UnsupportedError();
            };
        }
        for (const [query,] of this._iterateFunctions(this._manifest, 'queries')) {
            this._loaded.prototype['get_' + query] = function () {
                throw new errors_1.UnsupportedError();
            };
            this._loaded.prototype['subscribe_' + query] = function () {
                throw new errors_1.UnsupportedError();
            };
        }
        this._loaded.metadata = compat_1.makeBaseDeviceMetadata(manifest);
    }
    clearCache() {
        // nothing to do here
    }
    getDeviceClass() {
        return Promise.resolve(this._loaded);
    }
}
exports.default = UnsupportedBuiltinModule;
