// genie-toolkit/dist/lib/templates/wikidata.genie.out.js
"use strict";
// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2022 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Silei Xu <silei@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.$load = exports.boolean_question = exports.two_clean_filter_table = exports.two_with_filter_table = exports.one_clean_filter_table = exports.one_with_filter_table = void 0;
const $runtime = __importStar(require("../sentence-generator/runtime"));
const C = __importStar(require("./ast_manip"));
exports.one_with_filter_table = "one_with_filter_table";
exports.one_clean_filter_table = "one_clean_filter_table";
exports.two_with_filter_table = "two_with_filter_table";
exports.two_clean_filter_table = "two_clean_filter_table";
exports.boolean_question = "boolean_question";
/**
 * Load the templates in this file.
 *
 * This is an autogenerated method that will be called by the {@link SentenceGenerator}
 * when the templates are imported.
 */
async function $load($options, $locale, $grammar, $loader) {
    $grammar.declareSymbol("one_with_filter_table");
    if ($options.flags.wikidata) {
        $grammar.addRule("one_with_filter_table", [new $runtime.NonTerminal("complete_table", "table", undefined), new $runtime.NonTerminal("npp_id_soft_filter", "filter", undefined)], $locale._("${table} {with|having} ${filter}"), ((table, filter) => C.addFilter($loader, table, filter)), C.expressionKeyFn, { weight: (0.1) });
    }
    $grammar.declareSymbol("one_clean_filter_table");
    if ($options.flags.wikidata) {
        $grammar.addRule("one_clean_filter_table", [new $runtime.NonTerminal("complete_table", "table", undefined), new $runtime.NonTerminal("pvp_id_soft_filter", "filter", undefined)], $locale._("${table} ${filter}"), ((table, filter) => C.addFilter($loader, table, filter)), C.expressionKeyFn, { weight: (0.1) });
    }
    $grammar.declareSymbol("two_with_filter_table");
    if ($options.flags.wikidata) {
        $grammar.addRule("two_with_filter_table", [new $runtime.NonTerminal("one_with_filter_table", "table", undefined), new $runtime.NonTerminal("npp_id_soft_filter", "filter", undefined)], $locale._("${table} and ${filter}"), ((table, filter) => C.addFilter($loader, table, filter)), C.expressionKeyFn, { weight: (0.1) });
        $grammar.addRule("two_with_filter_table", [new $runtime.NonTerminal("one_which_filter_table", "table", undefined), new $runtime.NonTerminal("npp_id_soft_filter", "filter", undefined)], $locale._("${table} {and having|and with|, with|and have} ${filter}"), ((table, filter) => C.addFilter($loader, table, filter)), C.expressionKeyFn, { weight: (0.1) });
        $grammar.addRule("two_with_filter_table", [new $runtime.NonTerminal("one_clean_filter_table", "table", undefined), new $runtime.NonTerminal("npp_id_soft_filter", "filter", undefined)], $locale._("${table} {with|having} ${filter}"), ((table, filter) => C.addFilter($loader, table, filter)), C.expressionKeyFn, { weight: (0.1) });
    }
    $grammar.declareSymbol("two_clean_filter_table");
    if ($options.flags.wikidata) {
        $grammar.addRule("two_clean_filter_table", [new $runtime.NonTerminal("one_clean_filter_table", "table", undefined), new $runtime.NonTerminal("pvp_id_soft_filter", "filter", undefined)], $locale._("${table} ${filter}"), ((table, filter) => C.addFilter($loader, table, filter)), C.expressionKeyFn, { weight: (0.1) });
    }
    $grammar.declareSymbol("boolean_question");
    if ($options.flags.wikidata) {
        $grammar.addRule("boolean_question", [new $runtime.NonTerminal("complete_table", "table", undefined), new $runtime.NonTerminal("pvp_id_soft_filter", "filter", undefined)], $locale._("{are|were|is|was} there {a|} ${table} ${filter}"), ((table, filter) => C.makeVerificationQuestion($loader, table, filter)), undefined, { weight: (0.2) });
        $grammar.addRule("boolean_question", [new $runtime.NonTerminal("complete_table", "table", undefined), new $runtime.NonTerminal("npp_id_soft_filter", "filter", undefined)], $locale._("{are|were|is|was} there {a|} ${table} {with|having} ${filter}"), ((table, filter) => C.makeVerificationQuestion($loader, table, filter)), undefined, { weight: (0.2) });
        $grammar.addRule("boolean_question", [new $runtime.NonTerminal("with_filtered_table", "table", undefined), new $runtime.NonTerminal("pvp_id_soft_filter", "filter", undefined)], $locale._("{are|were|is|was} there {a|} ${table} ${filter}"), ((table, filter) => C.makeVerificationQuestion($loader, table, filter)), undefined, { weight: (0.1) });
        $grammar.addRule("boolean_question", [new $runtime.NonTerminal("with_filtered_table", "table", undefined), new $runtime.NonTerminal("npp_id_soft_filter", "filter", undefined)], $locale._("{are|were|is|was} there {a|} ${table} {with|having} ${filter}"), ((table, filter) => C.makeVerificationQuestion($loader, table, filter)), undefined, { weight: (0.1) });
    }
}
exports.$load = $load;
;
