// genie-toolkit/dist/lib/dialogue-agent/dialogue_queue.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2020 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
Object.defineProperty(exports, "__esModule", { value: true });
class QueueItem {
}
(function (QueueItem) {
    class UserInput extends QueueItem {
        constructor(command) {
            super();
            this.command = command;
        }
        toString() {
            return `UserInput(${this.command})`;
        }
    }
    QueueItem.UserInput = UserInput;
    class Notification extends QueueItem {
        constructor(app, outputType, outputValue) {
            super();
            this.app = app;
            this.outputType = outputType;
            this.outputValue = outputValue;
        }
        toString() {
            return `Notification(${this.app.uniqueId}, ${this.outputType})`;
        }
    }
    QueueItem.Notification = Notification;
    class Error extends QueueItem {
        constructor(app, error) {
            super();
            this.app = app;
            this.error = error;
        }
        toString() {
            return `Error(${this.app.uniqueId}, ${this.error})`;
        }
    }
    QueueItem.Error = Error;
})(QueueItem || (QueueItem = {}));
exports.default = QueueItem;
