// genie-toolkit/dist/lib/dialogue-agent/statement_executor.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Genie
//
// Copyright 2020 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const thingpedia_1 = require("thingpedia");
const thingtalk_1 = require("thingtalk");
// above MORE_SIZE, we set the "more" bit
const MORE_SIZE = 50;
// above PAGE_SIZE, we set the count but don't actually show the full list of results
const PAGE_SIZE = 10;
/**
 * Run the dialogue, executing ThingTalk and invoking the policy at the
 * right time.
 */
class InferenceStatementExecutor {
    constructor(engine, conversationId) {
        this._engine = engine;
        this._conversationId = conversationId;
    }
    _getType(schema, key) {
        if (key === '__device')
            return new thingtalk_1.Type.Entity('tt:device_id');
        const type = schema.getArgType(key);
        if (type)
            return type;
        if (key === 'distance') // HACK
            return new thingtalk_1.Type.Measure('m');
        return null;
    }
    _mapValue(key, type, value) {
        try {
            if (Number.isNaN(value) || (value instanceof Date && Number.isNaN(value.getTime())))
                throw new thingpedia_1.ImplementationError(`Invalid NaN value`);
            if (type instanceof thingtalk_1.Type.Enum && (typeof value !== 'string' || !type.entries.includes(value)))
                throw new thingpedia_1.ImplementationError(`Invalid enum value`);
            return thingtalk_1.Ast.Value.fromJS(type, value);
        }
        catch (e) {
            console.error(`Failed to map result field ${key} to type ${type}`, e);
            return null;
        }
    }
    async mapResult(schema, outputValue) {
        const mappedResult = {};
        for (const key in outputValue) {
            const value = outputValue[key];
            if (value === null || value === undefined)
                continue;
            if (key === '__timestamp')
                continue;
            const type = this._getType(schema, key);
            if (!type) {
                console.error(`Thingpedia function returned undeclared field ${key}`);
                continue;
            }
            let mapped;
            if (type instanceof thingtalk_1.Type.Compound)
                mapped = this._mapCompound(key + '.', schema, value);
            else
                mapped = this._mapValue(key, type, value);
            if (mapped)
                mappedResult[key] = mapped;
        }
        return new thingtalk_1.Ast.DialogueHistoryResultItem(null, mappedResult, outputValue);
    }
    _mapCompound(prefix, schema, object) {
        const result = {};
        for (const key in object) {
            const value = object[key];
            if (value === null || value === undefined)
                continue;
            const type = this._getType(schema, prefix + key);
            if (!type)
                continue;
            let mapped;
            if (type instanceof thingtalk_1.Type.Compound)
                mapped = this._mapCompound(prefix + key + '.', schema, object);
            else
                mapped = this._mapValue(prefix + key, type, value);
            if (mapped)
                result[key] = mapped;
        }
        return new thingtalk_1.Ast.Value.Object(result);
    }
    mapError(error) {
        const err = error;
        if (typeof err.code === 'string')
            return new thingtalk_1.Ast.Value.Enum(err.code);
        return new thingtalk_1.Ast.Value.String(error.message);
    }
    async _iterateResults(app, schema, into, intoRaw) {
        let count = 0;
        if (app === null)
            return [false, undefined];
        let error;
        for await (const value of app.mainOutput) {
            if (count >= MORE_SIZE)
                return [true, error];
            if (value instanceof Error) {
                error = value;
            }
            else {
                const mapped = await this.mapResult(schema, value.outputValue);
                into.push(mapped);
                intoRaw.push([value.outputType, value.outputValue]);
                count++;
            }
        }
        // if we get here, we iterated all results from the app, so we can stop
        return [false, error];
    }
    async executeStatement(stmt, privateState, notifications) {
        const program = new thingtalk_1.Ast.Program(null, [], [], [stmt]);
        const app = await this._engine.createApp(program, { notifications, conversation: this._conversationId });
        // by now the statement must have been typechecked
        (0, assert_1.default)(stmt.expression.schema);
        const results = [];
        const rawResults = [];
        const [more, error] = await this._iterateResults(app, stmt.expression.schema, results, rawResults);
        const annotations = {};
        let errorValue;
        if (error) {
            if (typeof error.code === 'number')
                errorValue = new thingtalk_1.Ast.Value.Enum(`http_${error.code}`);
            else if (typeof error.code === 'string' && error.code)
                errorValue = new thingtalk_1.Ast.Value.Enum(error.code);
            else
                errorValue = new thingtalk_1.Ast.Value.String(error.message);
            annotations.error_detail = new thingtalk_1.Ast.Value.String(error.message);
            if (error.stack)
                annotations.error_stack = new thingtalk_1.Ast.Value.String(error.stack);
        }
        const resultList = new thingtalk_1.Ast.DialogueHistoryResultList(null, results.slice(0, PAGE_SIZE), new thingtalk_1.Ast.Value.Number(results.length), more, errorValue);
        const newProgramRecord = {
            uniqueId: app.uniqueId,
            name: app.name,
            code: program.prettyprint(),
            results: rawResults.map((r) => r[1]),
            errors: errorValue ? [errorValue.toJS()] : [],
            icon: app.icon,
        };
        return [resultList, rawResults, newProgramRecord, undefined, { impl: annotations }];
    }
}
exports.default = InferenceStatementExecutor;
