// thingpedia/dist/helpers/index.js
"use strict";
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Thingpedia
//
// Copyright 2016-2019 The Board of Trustees of the Leland Stanford Junior University
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemoryPreferences = exports.FilePreferences = exports.ObjectSet = exports.RefCounted = exports.Xml = exports.Rss = exports.PollingStream = exports.OAuth2 = exports.Http = exports.Content = void 0;
/**
 * A collection of useful helper libraries bundled with the SDK.
 *
 * It is recommended, although not required, to use the following libraries
 * instead of custom bundled libraries.
 *
 * @namespace
 */
const Content = __importStar(require("./content"));
exports.Content = Content;
const Http = __importStar(require("./http"));
exports.Http = Http;
const oauth2_1 = __importDefault(require("./oauth2"));
exports.OAuth2 = oauth2_1.default;
const polling_1 = __importDefault(require("./polling"));
exports.PollingStream = polling_1.default;
const Rss = __importStar(require("./rss"));
exports.Rss = Rss;
const Xml = __importStar(require("./xml"));
exports.Xml = Xml;
const ref_counted_1 = __importDefault(require("./ref_counted"));
exports.RefCounted = ref_counted_1.default;
const ObjectSet = __importStar(require("./object_set"));
exports.ObjectSet = ObjectSet;
const file_prefs_1 = __importDefault(require("./file_prefs"));
exports.FilePreferences = file_prefs_1.default;
const memory_prefs_1 = __importDefault(require("./memory_prefs"));
exports.MemoryPreferences = memory_prefs_1.default;
