// genie-toolkit/dist/lib/dialogue-agent/notifications/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThingpediaNotificationBackend = exports.StaticNotificationBackends = void 0;
const twilio_1 = __importDefault(require("./twilio"));
const email_1 = __importDefault(require("./email"));
exports.StaticNotificationBackends = {
    'twilio': twilio_1.default,
    'email': email_1.default,
};
/**
 * Helper class to adapt a Thingpedia device into a notification backend.
 */
class ThingpediaNotificationBackend {
    constructor(device) {
        this.name = device.name;
        this.uniqueId = 'thingpedia/' + device.uniqueId;
        this._iface = device.queryInterface('notifications');
    }
    get requiredSettings() {
        return {};
    }
    notify(data) {
        return this._iface.notify(data);
    }
    notifyError(data) {
        return this._iface.notifyError(data);
    }
}
exports.ThingpediaNotificationBackend = ThingpediaNotificationBackend;
